README.md
## catbox-couchbase

[![NPM][npm-image]][npm-url] [![Build Status][travis-image]][travis-url]

Couchbase adapter for [catbox](https://github.com/hapijs/catbox).

### Installation

```
npm install catbox-couchbase
```

### Usage

#### Options

* `location` - Couchbase cluster location. Default `localhost:8091`
* `bucket` - Couchbase bucket. Default: `default`
* `partition` - Cache key prefix. Partition is intentionally used for cache key
  prefix since Couchbase recommends <= 10 buckets for per cluster.
  Also Catbox sets partition to `catbox` if it is not defined so use `NONE` for
  no partition.
* `flags` - Flags
  * `keepAlive` - Keep connection alive. Default `true`
  * `bypassCacheOnConnError` - Bypass cache if there is a connection error. Default `false`
  * `debug` - List of flags (`['events']`) for debugging. Default `[]`

*See [catbox client](https://github.com/hapijs/catbox#client) and 
[couchbase](https://github.com/couchbase/couchnode) for more details and options.*

### License

Licensed under The MIT License (MIT)  
For the full copyright and license information, please view the LICENSE.txt file.

[npm-url]: http://npmjs.org/package/catbox-couchbase
[npm-image]: https://badge.fury.io/js/catbox-couchbase.svg

[travis-url]: https://travis-ci.org/yieldbot/catbox-couchbase
[travis-image]: https://travis-ci.org/yieldbot/catbox-couchbase.svg?branch=master

CHANGELOG.md
## Changelog

### 2.2.4 (2016-01-06)

* Pin couchbase dep version

### 2.2.3 (2015-11-04)

* Upgrade couchbase to 2.1.x
* Upgrade dev dependencies
* Change license copyright

### 2.2.2 (2015-08-05)

* Update timeouts and keep alive

### 2.2.1 (2015-08-03)

* Change error event message

### 2.2.0 (2015-07-30)

* Add keep alive feature
* Add bypass event

### 2.1.0 (2015-07-28)

* Add bypass cache flag
* Add new server options

### 2.0.2 (2015-07-26)

* Misc.updates

### 2.0.1 (2015-07-18)

* Tidy changes

### 2.0.0 (2015-07-13)

* Use mock for tests
* Add bucket option
* Change behavior for partition option

### 1.0.0 (2015-07-12)

* Initial release

test/test-lib.js
/* jslint node: true */
/* global describe: false, it: false */
'use strict';

var Client = require('../'),
    expect = require('chai').expect;

// Tests

describe('lib', function() {

  var client;
  beforeEach(function() {
    client = new Client({flags: {isMock: true}});
  });
  afterEach(function() {
    client.stop();
  });

  it('should create a connection', function(done) {
    client.start(function(err) {
      expect(err).to.equal(undefined);
      expect(client.isReady()).to.equal(true);
      done();
    });
  });

  it('should end the connection', function(done) {
    client.start(function(err) {
      expect(err).to.equal(undefined);
      expect(client.isReady()).to.equal(true);
      client.stop();
      expect(client.isReady()).to.equal(false);
      done();
    });
  });

  it('should set and get an item', function(done) {
    client.start(function(err) {
      expect(err).to.equal(undefined);
      expect(client.isReady()).to.equal(true);

      var key = {id: 'foo', segment: 'test'};
      client.set(key, 'bar', 1000, function(err) {
        expect(err).to.equal(undefined);

        client.get(key, function(err, result) {
          expect(err).to.equal(null);
          expect(result).to.be.a('object');
          expect(result.item).to.equal('bar');
          done();
        });
      });
    });
  });

  it('should set and get an item (30 days)', function(done) {
    client.start(function(err) {
      expect(err).to.equal(undefined);
      expect(client.isReady()).to.equal(true);

      var key = {id: 'foo30', segment: 'test'};
      client.drop(key, function(err) {
        expect(err).to.equal(undefined);

        // TODO: Watch the issue
        // https://github.com/couchbase/couchnode/pull/41
        client.set(key, 'bar', 1000*60*60*24*29, function(err) {
          expect(err).to.equal(undefined);

          client.get(key, function(err, result) {
            expect(err).to.equal(null);
            expect(result).to.be.a('object');
            expect(result.item).to.equal('bar');
            done();
          });
        });
      });
    });
  });

  it('should set and get an item (>30 days)', function(done) {
    client.start(function(err) {
      expect(err).to.equal(undefined);
      expect(client.isReady()).to.equal(true);

      var key = {id: 'foo31', segment: 'test'};
      client.drop(key, function(err) {
        expect(err).to.equal(undefined);

        client.set(key, 'bar', 1000*60*60*24*31, function(err) {
          expect(err).to.equal(undefined);

          client.get(key, function(err, result) {
            expect(err).to.equal(null);
            expect(result).to.be.a('object');
            expect(result.item).to.equal('bar');
            done();
          });
        });
      });
    });
  });

  it('should fail to set an item due to value', function(done) {
    client.start(function(err) {
      expect(err).to.equal(undefined);
      expect(client.isReady()).to.equal(true);

      var key = {id: 'bar', segment: 'test'},
          val = {'baz': 'qux'};
      val.val = val;
      client.set(key, val, 1000, function(err) {
        expect(err.message).to.equal('Converting circular structure to JSON');
        done();
      });
    });
  });

  it('should fail to set an item due to key', function(done) {
    client.start(function(err) {
      expect(err).to.equal(undefined);
      expect(client.isReady()).to.equal(true);

      client.set(null, {}, 1000, function(err) {
        expect(err.message).to.equal('invalid key');

        client.set({}, {}, 1000, function(err) {
          expect(err.message).to.equal('invalid key');
          done();
        });
      });
    });
  });

  it('should fail to set an item due to ttl', function(done) {
    client.start(function(err) {
      expect(err).to.equal(undefined);
      expect(client.isReady()).to.equal(true);

      client.set({id: 'baz', segment: 'test'}, {}, 'boom', function(err) {
        expect(err.message).to.equal('invalid ttl');
        done();
      });
    });
  });

  it('should return null - key', function(done) {
    client.start(function(err) {
      expect(err).to.equal(undefined);
      expect(client.isReady()).to.equal(true);

      client.get(null, function(err, result) {
        expect(err).to.equal(null);
        expect(result).to.equal(null);
        done();
      });
    });
  });

  it('should fail to return - key', function(done) {
    client.start(function(err) {
      expect(err).to.equal(undefined);
      expect(client.isReady()).to.equal(true);

      client.get({}, function(err, result) {
        expect(err.message).to.equal('invalid key');
        expect(result).to.equal(undefined);
        done();
      });
    });
  });

  it('should drop an item', function(done) {
    client.start(function(err) {
      expect(err).to.equal(undefined);
      expect(client.isReady()).to.equal(true);

      var key = {id: 'qux', segment: 'test'};
      client.set(key, 'foo', 1, function(err) {
        expect(err).to.equal(undefined);
        client.drop(key, function(err) {
          expect(err).to.equal(undefined);
          client.get(key, function(err, result) {
            expect(err).to.equal(null);
            expect(result).to.equal(null);
            done();
          });
        });
      });
    });
  });

  it('should fail to drop an item due to key', function(done) {
    client.start(function(err) {
      expect(err).to.equal(undefined);
      expect(client.isReady()).to.equal(true);

      client.drop(null, function(err, result) {
        expect(err.message).to.equal('invalid key');
        expect(result).to.equal(undefined);

        client.drop({}, function(err, result) {
          expect(err.message).to.equal('invalid key');
          expect(result).to.equal(undefined);
          done();
        });
      });
    });
  });

  it('should validate segment', function(done) {
    expect(client.validateSegmentName().message).to.equal('invalid segment name');
    expect(client.validateSegmentName('\n').message).to.equal('invalid segment name');
    expect(client.validateSegmentName('\0').message).to.equal('invalid segment name');
    expect(client.validateSegmentName('\t').message).to.equal('invalid segment name');
    expect(client.validateSegmentName(' ').message).to.equal('invalid segment name');
    done();
  });

  it('should return null - ttl', function(done) {
    client.start(function(err) {
      expect(err).to.equal(undefined);
      expect(client.isReady()).to.equal(true);

      var key = {id: 'foobar', segment: 'test'};
      client.set(key, 'boom', 1, function(err) {
        expect(err).to.equal(undefined);
        setTimeout(function() {
          client.get(key, function(err, result) {
            expect(err).to.equal(null);
            expect(result).to.equal(null);
            done();
          });
        }, 2000);
      });
    });
  });

});

test/test-flags.js
/* jslint node: true */
/* global describe: false, it: false */
'use strict';

var Client = require('../'),
    expect = require('chai').expect;

// Tests

describe('flags', function() {

  var client;
  beforeEach(function() {
    client = new Client({location: '127.0.0.1:8092', flags: {bypassCacheOnConnError: true}});
  });
  afterEach(function() {
    client.stop();
  });

  it('should not fail', function(done) {
    client.start(function(err) {
      expect(err).to.equal(undefined);
      expect(client.isReady()).to.equal(true);

      var key = {id: 'foo', segment: 'test'};
      client.set(key, 'bar', 1000, function(err) {
        expect(err).to.equal(undefined);

        client.get(key, function(err, result) {
          expect(err).to.equal(null);
          expect(result).to.equal(null);

          client.drop(key, function(err) {
            expect(err).to.equal(undefined);
            done();
          });
        });
      });
    });
  });

});

test/test-catbox.js
/* jslint node: true */
/* global describe: false, it: false */
'use strict';

var Client = require('../'),
    Catbox = require('catbox'),
    expect = require('chai').expect;

// Tests

describe('catbox', function() {

  var client;
  beforeEach(function() {
    client = new Catbox.Client(Client, {flags: {isMock: true}});
  });
  afterEach(function() {
    client.stop();
  });

  it('should create a connection', function(done) {
    client.start(function(err) {
      expect(err).to.equal(undefined);
      expect(client.isReady()).to.equal(true);
      done();
    });
  });

  it('should end the connection', function(done) {
    client.start(function(err) {
      expect(err).to.equal(undefined);
      expect(client.isReady()).to.equal(true);
      client.stop();
      expect(client.isReady()).to.equal(false);
      done();
    });
  });

  it('should set and get an item', function(done) {
    client.start(function(err) {
      expect(err).to.equal(undefined);
      expect(client.isReady()).to.equal(true);

      var key = {id: 'foo', segment: 'test'};
      client.set(key, 'bar', 1000, function(err) {
        expect(err).to.equal(undefined);

        client.get(key, function(err, result) {
          expect(err).to.equal(null);
          expect(result).to.be.a('object');
          expect(result.item).to.equal('bar');
          done();
        });
      });
    });
  });

  it('should set and get an item (30 days)', function(done) {
    client.start(function(err) {
      expect(err).to.equal(undefined);
      expect(client.isReady()).to.equal(true);

      var key = {id: 'foo30', segment: 'test'};
      client.drop(key, function(err) {
        expect(err).to.equal(undefined);

        // TODO: Watch the issue
        // https://github.com/couchbase/couchnode/pull/41
        client.set(key, 'bar', 1000*60*60*24*29, function(err) {
          expect(err).to.equal(undefined);

          client.get(key, function(err, result) {
            expect(err).to.equal(null);
            expect(result).to.be.a('object');
            expect(result.item).to.equal('bar');
            done();
          });
        });
      });
    });
  });

  it('should set and get an item (>30 days)', function(done) {
    client.start(function(err) {
      expect(err).to.equal(undefined);
      expect(client.isReady()).to.equal(true);

      var key = {id: 'foo31', segment: 'test'};
      client.drop(key, function(err) {
        expect(err).to.equal(undefined);

        client.set(key, 'bar', 1000*60*60*24*31, function(err) {
          expect(err).to.equal(undefined);

          client.get(key, function(err, result) {
            expect(err).to.equal(null);
            expect(result).to.be.a('object');
            expect(result.item).to.equal('bar');
            done();
          });
        });
      });
    });
  });

  it('should fail to set an item due to value', function(done) {
    client.start(function(err) {
      expect(err).to.equal(undefined);
      expect(client.isReady()).to.equal(true);

      var key = {id: 'bar', segment: 'test'},
          val = {'baz': 'qux'};
      val.val = val;
      client.set(key, val, 1000, function(err) {
        expect(err.message).to.equal('Converting circular structure to JSON');
        done();
      });
    });
  });

  it('should fail to set an item due to key', function(done) {
    client.start(function(err) {
      expect(err).to.equal(undefined);
      expect(client.isReady()).to.equal(true);

      client.set(null, {}, 1000, function(err) {
        expect(err.message).to.equal('Invalid key');

        client.set({}, {}, 1000, function(err) {
          expect(err.message).to.equal('Invalid key');
          done();
        });
      });
    });
  });

  it('should fail to set an item due to ttl', function(done) {
    client.start(function(err) {
      expect(err).to.equal(undefined);
      expect(client.isReady()).to.equal(true);

      client.set({id: 'baz', segment: 'test'}, {}, 'boom', function(err) {
        expect(err.message).to.equal('invalid ttl');
        done();
      });
    });
  });

  it('should return null - key', function(done) {
    client.start(function(err) {
      expect(err).to.equal(undefined);
      expect(client.isReady()).to.equal(true);

      client.get(null, function(err, result) {
        expect(err).to.equal(null);
        expect(result).to.equal(null);
        done();
      });
    });
  });

  it('should fail to return - key', function(done) {
    client.start(function(err) {
      expect(err).to.equal(undefined);
      expect(client.isReady()).to.equal(true);

      client.get({}, function(err, result) {
        expect(err.message).to.equal('Invalid key');
        expect(result).to.equal(undefined);
        done();
      });
    });
  });

  it('should drop an item', function(done) {
    client.start(function(err) {
      expect(err).to.equal(undefined);
      expect(client.isReady()).to.equal(true);

      var key = {id: 'qux', segment: 'test'};
      client.set(key, 'foo', 1, function(err) {
        expect(err).to.equal(undefined);
        client.drop(key, function(err) {
          expect(err).to.equal(undefined);
          client.get(key, function(err, result) {
            expect(err).to.equal(null);
            expect(result).to.equal(null);
            done();
          });
        });
      });
    });
  });

  it('should fail to drop an item due to key', function(done) {
    client.start(function(err) {
      expect(err).to.equal(undefined);
      expect(client.isReady()).to.equal(true);

      client.drop(null, function(err, result) {
        expect(err.message).to.equal('Invalid key');
        expect(result).to.equal(undefined);

        client.drop({}, function(err, result) {
          expect(err.message).to.equal('Invalid key');
          expect(result).to.equal(undefined);
          done();
        });
      });
    });
  });

  it('should validate segment', function(done) {
    expect(client.validateSegmentName().message).to.equal('invalid segment name');
    expect(client.validateSegmentName('\n').message).to.equal('invalid segment name');
    expect(client.validateSegmentName('\0').message).to.equal('invalid segment name');
    expect(client.validateSegmentName('\t').message).to.equal('invalid segment name');
    expect(client.validateSegmentName(' ').message).to.equal('invalid segment name');
    done();
  });

  it('should return null - ttl', function(done) {
    client.start(function(err) {
      expect(err).to.equal(undefined);
      expect(client.isReady()).to.equal(true);

      var key = {id: 'foobar', segment: 'test'};
      client.set(key, 'boom', 1, function(err) {
        expect(err).to.equal(undefined);
        setTimeout(function() {
          client.get(key, function(err, result) {
            expect(err).to.equal(null);
            expect(result).to.equal(null);
            done();
          });
        }, 2000);
      });
    });
  });

});

lib/index.js
/*
 * catbox-couchbase
 * Copyright (c) 2015 Yieldbot, Inc. (http://github.com/yieldbot/catbox-couchbase)
 * For the full copyright and license information, please view the LICENSE.txt file.
 */

/* jslint node: true */
'use strict';

var couchbase    = require('couchbase'),
    util         = require('util'),
    EventEmitter = require('events').EventEmitter;

// Init the module
module.exports = function(options) {

  var serverOptions    = options || {},
      serverLocations  = '127.0.0.1:8091',
      serverBucket     = 'default',
      serverPartition, // cache key prefix
      eEmitter,        // event emitter
      flags,           // flags (e.g. `{flags: {debug: ['events']}}`)
      cacheClient,     // cache client
      cluster;         // cluster object

  var TTL_MAX = 1000*60*60*24*30; // 2592000000 milliseconds = 30 days

  // Initialize event emitter
  var EEmitter = function EEmitter() { EventEmitter.call(this); };
  util.inherits(EEmitter, EventEmitter);
  eEmitter = new EEmitter();
  eEmitter.on('error', function() { }); // prevent unhandled error event

  // Server options
  // REF: http://docs.couchbase.com/sdk-api/couchbase-node-client-2.0.8/index.html
  //      http://docs.couchbase.com/developer/dev-guide-3.0/doc-expiration.html

  if(serverOptions.location) {
    serverLocations = (serverOptions.location instanceof Array) ? serverOptions.location.join(',') : ''+serverOptions.location;
  }

  if(typeof serverOptions.partition === 'string' && serverOptions.partition !== 'NONE') {
    serverPartition = serverOptions.partition;
  }

  if(typeof serverOptions.bucket === 'string' && serverOptions.bucket) {
    serverBucket = serverOptions.bucket;
  }

  if(!serverOptions.connectionTimeout) serverOptions.connectionTimeout = 60000; // default 5000
  if(!serverOptions.operationTimeout)  serverOptions.operationTimeout  = 50000; // default 2500

  // Flags
  flags = (serverOptions.flags && typeof serverOptions.flags === 'object' && serverOptions.flags) || {};

  var bypassCacheOnConnError = !!flags.bypassCacheOnConnError,
      keepAliveFlag          = (flags.keepAlive === false) ? false : true,
      keepAliveTimer;

  // If mock flag is true then
  if(flags.isMock) {
    couchbase = require('couchbase').Mock; // use mock
  }

  // Init cluster
  // TODO: Watch the issue
  // https://issues.couchbase.com/browse/JSCBC-156
  // var cnstrQueryParams = [
  //   util.format('connectionTimeout=%s', serverOptions.connectionTimeout),
  //   util.format('operationTimeout=%s', serverOptions.operationTimeout)
  // ];
  // var cnstr = util.format('couchbase://%s/?%s', serverLocations, cnstrQueryParams.join('&'));
  cluster = new couchbase.Cluster(util.format('couchbase://%s/', serverLocations));

  // Creates a connection to the cache server
  var start = function start(callback) {

    if(typeof callback !== 'function') {
      throw new Error('invalid callback function');
    }

    // If there is an instance then
    if(cacheClient) {
      return callback();
    }

    // Create instance
    cacheClient = openBucket();

    // Keep alive
    if(keepAliveFlag) {
      keepAlive();
    }

    // Set events
    eEmitter.emit('start', util.format('started for %s/pools/default/buckets/%s', serverLocations, serverBucket));

    return callback();
  };

  // Ends the cache server connection
  var stop = function stop() {
    if(cacheClient) {
      cacheClient.disconnect(); // end connection
      cacheClient = null;       // cleanup
      cluster     = null;       // cleanup

      // Clear keep alive interval
      if(keepAliveTimer) {
        clearInterval(keepAliveTimer);
      }

      eEmitter.emit('stop', util.format('stopped for %s/pools/default/buckets/%s', serverLocations, serverBucket));
    }
  };

  // Returns whether the cache server is ready or not
  var isReady = function isReady() {
    return (!!cacheClient);
  };

  // Validates segment name
  var validateSegmentName = function validateSegmentName(name) {
    if(!name || name.indexOf('\n') !== -1 || name.indexOf('\0') !== -1 || name.indexOf('\t') !== -1 || name.indexOf(' ') !== -1) {
      return new Error('invalid segment name');
    }
    return null; // catbox wants null
  };

  // Gets a value for the given key
  var get = function get(key, callback) {

    if(typeof callback !== 'function') {
      throw new Error('invalid callback function');
    }
    else if(!isReady()) {
      return callback(new Error('cache server is not ready'));
    }

    // If key is falsy then
    if(!key) {
      return callback(null, null); // return null like catbox
    }

    // Get cache value
    var keyF = generateKey(key);

    if(!keyF) {
      return callback(new Error('invalid key'));
    }

    eEmitter.emit('get', keyF);

    if(!cacheClient.connected && bypassCacheOnConnError) {
      eEmitter.emit('bypass', 'get '+keyF+' (c)');
      return callback(null, null); // bypass cache
    }

    cacheClient.get(keyF, function(err, result) {
      if(err) {
        // Code 16: Generic network failure.
        if(bypassCacheOnConnError && err.code === 16) {
          eEmitter.emit('bypass', 'get '+keyF+' (g)');
          return callback(null, null); // bypass cache
        }
        // If error code is not 13 (The key does not exist on the server) then
        if(err.code !== 13) {
          return callback(err);
        }
      }

      // No result then no cache
      if(!result || !result.value) {
        return callback(null, null);
      }

      // Parse cache content
      var envelope = null;
      try {
        envelope = JSON.parse(result.value);
      }
      catch(e) {
        eEmitter.emit('error', e);
        return callback(e);
      }

      // Check cache content
      if(!envelope || !envelope.item || !envelope.stored) {
        return callback(new Error('invalid cache content'));
      }

      return callback(null, envelope);
    });
  };

  // Sets a value for the given key
  var set = function set(key, value, ttl, callback) {

    if(typeof callback !== 'function') {
      throw new Error('invalid callback function');
    }
    else if(!key) {
      return callback(new Error('invalid key'));
    }
    else if(typeof ttl !== 'number') {
      return callback(new Error('invalid ttl'));
    }
    else if(!isReady()) {
      return callback(new Error('cache server is not ready'));
    }

    // Prepare cache content
    var envelope = {item: value, stored: Date.now(), ttl: ttl},
        envelopeStr;

    try {
      envelopeStr = JSON.stringify(envelope);
    }
    catch(e) {
      eEmitter.emit('error', e);
      return callback(e);
    }

    // Set cache
    var keyF = generateKey(key);

    if(!keyF) {
      return callback(new Error('invalid key'));
    }

    eEmitter.emit('set', keyF);

    if(!cacheClient.connected && bypassCacheOnConnError) {
      eEmitter.emit('bypass', 'set '+keyF+' (c)');
      return callback(); // bypass cache
    }

    // Expiry
    var expiry = Math.max(1, Math.floor(envelope.ttl / 1000)); // min is 1 sec
    // If ttl is greater than max ttl
    if(envelope.ttl > TTL_MAX) {
      // REF: http://docs.couchbase.com/developer/dev-guide-3.0/doc-expiration.html
      // Calculate timestamp
      expiry = Math.max(1, Math.floor((new Date(envelope.stored + envelope.ttl).getTime()) / 1000)); // min is 1 sec
    }

    cacheClient.upsert(keyF, envelopeStr, {expiry: expiry}, function(err) {
      if(err) {
        // Code 16: failed to connect to bucket.
        if(bypassCacheOnConnError && err.code === 16) {
          eEmitter.emit('bypass', 'set '+keyF+' (u)');
          return callback(); // bypass cache
        }
        return callback(err);
      }

      return callback();
    });
  };

  // Remove the given key
  var drop = function drop(key, callback) {

    if(typeof callback !== 'function') {
      throw new Error('invalid callback function');
    }
    else if(!isReady()) {
      return callback(new Error('cache server is not ready'));
    }

    // Drop cache
    var keyF = generateKey(key);

    if(!keyF) {
      return callback(new Error('invalid key'));
    }

    eEmitter.emit('drop', keyF);

    if(!cacheClient.connected && bypassCacheOnConnError) {
      eEmitter.emit('bypass', 'drop '+keyF+' (c)');
      return callback(); // bypass cache
    }

    cacheClient.remove(keyF, function(err) {
      if(err) {
        // Code 16: failed to connect to bucket.
        if(bypassCacheOnConnError && err.code === 16) {
          eEmitter.emit('bypass', 'drop '+keyF+' (r)');
          return callback(); // bypass cache
        }
        // If error code is not 13 (The key does not exist on the server) then
        if(err.code !== 13) {
          return callback(err);
        }
      }

      return callback();
    });
  };

  // Generates keys
  var generateKey = function generateKey(key) {
    return (serverPartition ? encodeURI(serverPartition)+':' : '') + (key && key.segment ? encodeURI(key.segment) : '') + (key && key.id ? ':'+encodeURI(key.id) : '');
  };

  // Allows hook on events
  var on = function on(event, callback) {
    eEmitter.addListener(event, callback);
  };

  // Opens bucket
  var openBucket = function openBucket() {

    // NOTE: CB library creates new instance for each open bucket call.

    if(cacheClient && cacheClient.disconnected) {
      cacheClient.disconnected();
    }

    // Create instance
    cacheClient = cluster.openBucket(serverBucket);
    // Code 24: failed to connect to bucket

    // Set options
    cacheClient.connectionTimeout = serverOptions.connectionTimeout;
    cacheClient.operationTimeout  = serverOptions.operationTimeout;

    // Set events
    cacheClient.on('connect', function() {
      eEmitter.emit('connect', util.format('connected to %s bucket', serverBucket));
    });
    cacheClient.on('error', function(err) {
      eEmitter.emit('error', {code: err.code, message: err.message});
    });

    return cacheClient;
  };

  // Keeps connection alive
  var keepAlive = function keepAlive() {

    var running = false;

    keepAliveTimer = setInterval(function() {
      // If it is not running then
      if(!running) {
        running = true;

        // If it is not connected then
        if(!cacheClient.connected) {
          eEmitter.emit('keepalive', util.format('trying to connect... (connected=%s)', cacheClient.connected));
          cacheClient = openBucket();
          running = false;
        } else {
          // Get an item
          cacheClient.get('keepalive', function(err) {
            // Code 23: Client-Side timeout exceeded for operation
            // Code 16: Generic network failure
            // If error code is not 13 (The key does not exist on the server) then
            if(err && err.code !== 13) {
              eEmitter.emit('keepalive', util.format('trying to connect... (connected=%s err=%s)', cacheClient.connected, err.code));
              cacheClient = openBucket();
            }
            else {
              eEmitter.emit('keepalive', 'connection is alive');
            }
            running = false;
          });
        }
      }
    }, 2000);
  };

  // Debug
  //flags = {debug: ['events']}; // for debug
  if(flags.debug instanceof Array) {
    // If events in debug flags
    if(flags.debug.indexOf('events') !== -1) {
      // Log all the events
      ['start', 'stop', 'get', 'set', 'drop', 'connect', 'error', 'keepalive', 'bypass'].forEach(function(event) {
        on(event, function(data) {
          console.log('\nDebug: ' + JSON.stringify({event: event, data:data}) + '\n');
        });
      });
    }
  }

  // Return
  return {
    start:               start,
    stop:                stop,
    isReady:             isReady,
    validateSegmentName: validateSegmentName,
    get:                 get,
    set:                 set,
    drop:                drop,
    generateKey:         generateKey,
    on:                  on
  };
};